src/lib.rs

```diff
@@ -540,6 +605,57 @@ impl Assembler {
     );
 }
 
+/// Source and target for automatic program wrapping.
+///
+/// After the instruction at offset pointed by [`source`] has been executed, the program control flow jumps to the
+/// instruction pointed by [`target`]. If the instruction is a jump, and the condition is true, the jump takes priority.
+///
+/// [`source`]: Self::source
+/// [`target`]: Self::target
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Wrap {
+    /// Source instruction for wrap.
+    pub source: u8,
+    /// Target instruction for wrap.
+    pub target: u8,
+}
+
+/// Program ready to be executed by PIO hardware.
+#[derive(Debug)]
+pub struct Program<PublicDefines> {
+    /// Assembled program code.
+    pub code: ArrayVec<u16, MAX_PROGRAM_SIZE>,
+    /// Offset at which the program must be loaded.
+    ///
+    /// Most often 0 if defined. This might be needed when using data based `JMP`s.
+    pub origin: Option<u8>,
+    /// Wrapping behavior for this program.
+    pub wrap: Wrap,
+    /// Side-set info for this program.
+    pub side_set: SideSet,
+    /// Public defines for the program.
+    pub public_defines: PublicDefines,
```

Comment:
with this new separation i think it is clear these defines are a product of the parser, not of every pio program. I'd change the parser to return `{ defines, program }` instead.

Reply:
This makes sense, and I was quite unhappy with needing to use `Program<()>` instead of `Program` everywhere.

Reply:
This is now fixed. `pio_parser` defines its own `Program` type which wraps the `pio::Program` inside.